src/firefly/java/edu/caltech/ipac/table/io/FITSTableReader.java

```diff
@@ -393,9 +556,117 @@ private static DataGroup convertStarTableToDataGroup(StarTable table,
         }
         dataGroup.trimToSize();
         return dataGroup;
+    }*/
+
+
+    //This function is based on the packageValue function from the FitsStarTable class in the uk.ac.starlink.fits package
+    private static Object getValAsObject(Object base, int icol, boolean[] hasBlank, long[] blanks, boolean[] isScaled,
+                                         double[] scales, double[] zeros) {
+        if (base == null) {
+            return null;
+        } else {
+            Class bcls = base.getClass();
+            Class cls = bcls.isArray() ? bcls.getComponentType() : null;
+            if (cls != null && Array.getLength(base) == 1) {
+                boolean hasblank = hasBlank[icol];
+                long blank = blanks[icol];
+                boolean scaled = isScaled[icol];
+                double scale = scales[icol];
+                double zero = zeros[icol];
+                if (cls == Byte.TYPE) {
+                    byte val = ((byte[])(base))[0];
+                    /*
+                    TODO: Attempt to use ternary operator here instead
+                     */
+                    if (hasblank && val == (byte)((int)blank))
+                        return null;
+                    if (scaled)
+                        return Double.valueOf((double)val * scale + zero);
+                    else
+                        return Byte.valueOf(val);
```

Comment:
@robyww I made all of the other changes, but I tested this and think using the ternary operator here wouldn't work (because I am trying to give it two different return types). For the following line: `return scaled ? val * scale + zero : val;` - even is **scaled** is **false** and **val** is, say, **byte** or **int** or **short**, **val** will get cast to **double** implicitly if we use a ternary operator here (because `val * scale + zero` is **double**). So I had to leave the if/else as is. 

But removing the explicit class conversion and letting it autobox works as you suggested.

Reply:
That really surprises me. Have you tried calling the function in the debugger to confirm?

You also might try
```java
return scaled ? (double)(val * scale + zero) : (int)val;
```